lib/internal/modules/esm/assert.js

```diff
@@ -0,0 +1,104 @@
+'use strict';
+
+const {
+  ArrayPrototypeIncludes,
+  ArrayPrototypeFind,
+  ObjectCreate,
+  ObjectEntries,
+  ObjectValues,
+  ObjectPrototypeHasOwnProperty,
+  Symbol,
+} = primordials;
+
+const {
+  ERR_IMPORT_ASSERTION_TYPE_FAILED,
+  ERR_IMPORT_ASSERTION_TYPE_MISSING,
+  ERR_IMPORT_ASSERTION_TYPE_UNSUPPORTED,
+} = require('internal/errors').codes;
+
+const kImplicitAssertType = Symbol('implicit assert type');
+
+/**
+ * Define a map of module formats to import assertion types (the value of `type`
+ * in `assert { type: 'json' }`).
+ * @type {Map<string, string | typeof kImplicitAssertType}
+ */
+const formatTypeMap = {
+  '__proto__': null,
+  'builtin': kImplicitAssertType,
+  'commonjs': kImplicitAssertType,
+  'json': 'json',
+  'module': kImplicitAssertType,
+  'wasm': kImplicitAssertType, // Should probably be 'webassembly' per https://github.com/tc39/proposal-import-assertions
+};
+
+function findExpectedFormat(type) {
+  return ArrayPrototypeFind(
+    ObjectEntries(formatTypeMap),
+    ({ 1: value }) => value === type
+  )?.[0];
+}
+
+/**
+ * Test a module import's assertions.
+ * @param {string} format One of Node's supported translators
+ * @param {Record<string, string>} importAssertions The `"assert"` value passed
+ *                                                  to the `AssertClause` or
+ *                                                  the `optionsExpression`
+ *                                                  optional parameter of a a
+ *                                                  dynamic import, or an empty
+ *                                                  object if no value was
+ *                                                  provided.
+ * @param {string} [url] for error reporting.
+ * @returns {true}
+ * @throws {TypeError} if the format and import assertions are not compatible.
+ */
+function validateAssertions(format, importAssertions = ObjectCreate(null),
+                            url) {
+  const validType = formatTypeMap[format];
+
+  switch (validType) {
+    case undefined:
+      // Ignore assertions for module types we don't recognize, to allow new
+      // formats in the future.
+      return true;
+
+    case importAssertions.type:
+      // The asserted type is the valid type for this format.
+      return true;
+
+    case kImplicitAssertType:
+      // This format doesn't allow an import assertion type, so the property
+      // must not be set on the import assertions object.
+      if (!ObjectPrototypeHasOwnProperty(importAssertions, 'type')) {
+        return true;
+      }
+      if (!ArrayPrototypeIncludes(ObjectValues(formatTypeMap),
```

Comment:
this values array can be cached at module level, so it doesn't have to be recomputed each time